devmand/gateway/src/devmand/channels/cli/QueuedCli.cpp

```diff
@@ -15,16 +15,32 @@ using namespace folly;
 QueuedCli::QueuedCli(
     string _id,
     shared_ptr<Cli> _cli,
-    const shared_ptr<Executor>& _parentExecutor)
+    shared_ptr<Executor> _parentExecutor)
     : id(_id),
       cli(_cli),
+      parentExecutor(_parentExecutor),
       serialExecutorKeepAlive(SerialExecutor::create(
-          Executor::getKeepAliveToken(_parentExecutor.get()))) {}
+          Executor::getKeepAliveToken(_parentExecutor.get()))) {
+  isProcessing = false;
+  shutdown = false;
+}
 
 QueuedCli::~QueuedCli() {
-  // Execute a dummy task to make sure all running tasks have finished and that
-  // we can safely close this instance
-  via(serialExecutorKeepAlive, []() {}).get();
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~QCli";
+  shutdown = true;
+  isProcessing = true;
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~QCli: dequeuing " << queue.size() << " items";
+  QueueEntry queueEntry;
+  while (queue.try_dequeue(queueEntry)) {
+    queueEntry.promise->setException(runtime_error("QCli: Shutting down"));
+  }
+  serialExecutorKeepAlive = nullptr;
```

Comment:
I needed those for fixing a deadlock. Member would be destroyed after destructor, where it would just hang with no obvious reason and no easy way to debug.

Reply:
Is this still a problem after shared_from_this has been added ?